# 100 pandas puzzles

## Importing pandas

### Getting started and checking your pandas setup
Difficulty:easy

**1**. Import pandas under the alias pd.

In [1]:
import pandas as pd
import numpy as np

**2**. Print the version of pandas that has been imported.

In [3]:
print(pd.__version__)

2.2.3


**3**. Print out all the version information of the libraries that are required by the pandas library.

In [10]:
!pipdeptree -p pandas

pandas==2.2.3
├── numpy [required: >=1.26.0, installed: 2.2.2]
├── python-dateutil [required: >=2.8.2, installed: 2.9.0.post0]
│   └── six [required: >=1.5, installed: 1.17.0]
├── pytz [required: >=2020.1, installed: 2025.1]
└── tzdata [required: >=2022.7, installed: 2025.1]


## DataFrame basics

### A few of the fundamental routines for selecting, sorting, adding and aggregating data in DataFrames
Difficulty: easy

    Note: remember to import numpy using:

    import numpy as np
    Consider the following Python dictionary data and Python list labels:

    data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
            'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
            'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
            'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

    labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
    (This is just some meaningless data I made up with the theme of animals and trips to a vet.)

**4**. Create a DataFrame df from this dictionary data which has the index labels.

In [13]:
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

In [14]:
df = pd.DataFrame.from_dict(data, orient='index', columns=labels)
df

a    b      c    d    e    f      g    h    i    j
animal    cat  cat  snake  dog  dog  cat  snake  cat  dog  dog
age       2.5    3    0.5  NaN    5    2    4.5  NaN    7    3
visits      1    3      2    3    2    3      1    1    2    1
priority  yes  yes     no  yes   no   no     no  yes   no   no

**5**. Display a summary of the basic information about this DataFrame and its data (hint: there is a single method that can be called on the DataFrame).

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, animal to priority
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   a       4 non-null      object
 1   b       4 non-null      object
 2   c       4 non-null      object
 3   d       3 non-null      object
 4   e       4 non-null      object
 5   f       4 non-null      object
 6   g       4 non-null      object
 7   h       3 non-null      object
 8   i       4 non-null      object
 9   j       4 non-null      object
dtypes: object(10)
memory usage: 352.0+ bytes


In [330]:
df2 = pd.DataFrame(data, index=labels)
df2

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [199]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   animal    10 non-null     object 
 1   age       8 non-null      float64
 2   visits    10 non-null     int64  
 3   priority  10 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


**6**. Return the first 3 rows of the DataFrame df.

In [200]:
df2.head(3)

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

In [201]:
df2[0:3]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

**7**. Select just the 'animal' and 'age' columns from the DataFrame df.

In [232]:
df2[ ["animal", "age"] ]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  1.5
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [258]:
df2.loc[:,['animal', 'age']]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  1.5
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

**8**. Select the data in rows [3, 4, 8] and in columns ['animal', 'age'].

In [235]:
df2.loc[['c','d','h'],['animal', 'age']]

animal  age
c  snake  0.5
d    dog  NaN
h    cat  NaN

**9**. Select only the rows where the number of visits is greater than 3.

In [207]:
df2[ df2.visits > 3]

Empty DataFrame
Columns: [animal, age, visits, priority]
Index: []

In [208]:
df2[lambda x: x.visits > 3]

Empty DataFrame
Columns: [animal, age, visits, priority]
Index: []

**10**. Select the rows where the age is missing, i.e. it is NaN.

In [209]:
df2[ df2.age.isna()]

animal  age  visits priority
d    dog  NaN       3      yes
h    cat  NaN       1      yes

In [210]:
df2[ lambda x: x.age.isna() ]

animal  age  visits priority
d    dog  NaN       3      yes
h    cat  NaN       1      yes

**11**. Select the rows where the animal is a cat and the age is less than 3.

In [211]:
df2[ (df2.animal == 'cat') & (df2.age < 3) ]

animal  age  visits priority
a    cat  2.5       1      yes
f    cat  2.0       3       no

**12**. Select the rows the age is between 2 and 4 (inclusive).

In [212]:
df2[ (df2.age >=2) & (df2.age <=4) ]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
f    cat  2.0       3       no
j    dog  3.0       1       no

In [213]:
df2[ df2.age.between(2,4, inclusive='both') ]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
f    cat  2.0       3       no
j    dog  3.0       1       no

**13**. Change the age in row 'f' to 1.5.

In [252]:
df2.loc['f', 'age'] = 1.5
df2

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  1.5       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

**14**. Calculate the sum of all visits in df (i.e. find the total number of visits).

In [251]:
df2.visits.sum()

19

**15**. Calculate the mean age for each different animal in df.

In [253]:
df2.groupby('animal').age.mean()

animal
cat      2.333333
dog      5.000000
snake    2.500000
Name: age, dtype: float64

**16**. Append a new row 'k' to df with your choice of values for each column. Then delete that row to return the original DataFrame.

In [263]:
new_row = pd.DataFrame({'animal':'fig', 'age':10.0, 'visits':1, 'priority':'yes'}, index=['k'])
new_row

animal   age  visits priority
k    fig  10.0       1      yes

In [268]:
pd.concat([df2, new_row], ignore_index=False)

animal   age  visits priority
a    cat   2.5       1      yes
b    cat   3.0       3      yes
c  snake   0.5       2       no
d    dog   NaN       3      yes
e    dog   5.0       2       no
f    cat   1.5       3       no
g  snake   4.5       1       no
h    cat   NaN       1      yes
i    dog   7.0       2       no
j    dog   3.0       1       no
k    fig  10.0       1      yes

In [269]:
pd.concat([df2, new_row], ignore_index=False).drop('k')

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  1.5       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

**17**. Count the number of each type of animal in df.

In [283]:
df2.groupby('animal').animal.count()

animal
cat      4
dog      4
snake    2
Name: animal, dtype: int64

In [288]:
df2.groupby('animal').agg({'animal':'count'})

animal
animal        
cat          4
dog          4
snake        2

**18**. Sort df first by the values in the 'age' in decending order, then by the value in the 'visits' column in ascending order (so row i should be first, and row d should be last).

In [292]:
df2.sort_values('visits').sort_values(by='age', ascending=False)

animal  age  visits priority
i    dog  7.0       2       no
e    dog  5.0       2       no
g  snake  4.5       1       no
j    dog  3.0       1       no
b    cat  3.0       3      yes
a    cat  2.5       1      yes
f    cat  1.5       3       no
c  snake  0.5       2       no
h    cat  NaN       1      yes
d    dog  NaN       3      yes

In [294]:
df2.sort_values(by=['age', 'visits'], ascending=[False,True])

animal  age  visits priority
i    dog  7.0       2       no
e    dog  5.0       2       no
g  snake  4.5       1       no
j    dog  3.0       1       no
b    cat  3.0       3      yes
a    cat  2.5       1      yes
f    cat  1.5       3       no
c  snake  0.5       2       no
h    cat  NaN       1      yes
d    dog  NaN       3      yes

**19**. The 'priority' column contains the values 'yes' and 'no'. Replace this column with a column of boolean values: 'yes' should be True and 'no' should be False.

In [322]:
df2.priority = df2.priority.map( lambda x:
    x == 'yes'
)
df2

animal  age  visits  priority
a    cat  2.5       1      True
b    cat  3.0       3      True
c  snake  0.5       2     False
d    dog  NaN       3      True
e    dog  5.0       2     False
f    cat  2.0       3     False
g  snake  4.5       1     False
h    cat  NaN       1      True
i    dog  7.0       2     False
j    dog  3.0       1     False

**20**. In the 'animal' column, change the 'snake' entries to 'python'.

In [331]:
df2

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [332]:
def convAnimal(animal):
    if animal == 'snake':
        return 'python'
    else:
        return animal

df3 = df2.copy()

df3.animal = df2.animal.apply( lambda x:
    convAnimal(x)
)
df3

animal  age  visits priority
a     cat  2.5       1      yes
b     cat  3.0       3      yes
c  python  0.5       2       no
d     dog  NaN       3      yes
e     dog  5.0       2       no
f     cat  2.0       3       no
g  python  4.5       1       no
h     cat  NaN       1      yes
i     dog  7.0       2       no
j     dog  3.0       1       no

**21**. For each animal type and each number of visits, find the mean age. In other words, each row is an animal, each column is a number of visits and the values are the mean ages (hint: use a pivot table).

In [365]:
#  Multi-Index['animal','visits']
df2.groupby(['animal','visits']).agg({
    'age':'mean'
})

age
animal visits     
cat    1       2.5
       3       2.5
dog    1       3.0
       2       6.0
       3       NaN
snake  1       4.5
       2       0.5

In [366]:
# Multi-Index into column
df2.groupby(['animal','visits']).agg({
    'age':'mean'
}).reset_index()

animal  visits  age
0    cat       1  2.5
1    cat       3  2.5
2    dog       1  3.0
3    dog       2  6.0
4    dog       3  NaN
5  snake       1  4.5
6  snake       2  0.5

In [369]:
result = df2.groupby(['animal','visits']).agg({
        'age':'mean'
    }).reset_index() .pivot( index="animal", columns="visits", values="age")

result


visits    1    2    3
animal               
cat     2.5  NaN  2.5
dog     3.0  6.0  NaN
snake   4.5  0.5  NaN

## DataFrames: beyond the basics

### Slightly trickier: you may need to combine two or more methods to get the right answer
Difficulty:

    The previous section was tour through some basic but essential DataFrame operations.
    Below are some ways that you might need to cut your data, but for which there is no single "out of the box" method.

**22**. You have a DataFrame df with a column 'A' of integers. For example:

    df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
    How do you filter out rows which contain the same integer as the row immediately above?

    You should be left with a column containing the following values:

    1, 2, 3, 4, 5, 6, 7

In [390]:
df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
df.A.unique().tolist()

[1, 2, 3, 4, 5, 6, 7]

**23**. Given a DataFrame of numeric values, say

    df = pd.DataFrame(np.random.random(size=(5, 3))) # a 5x3 frame of float values
    how do you subtract the row mean from each element in the row?

In [391]:
df = pd.DataFrame(np.random.random(size=(5, 3)))
df

0         1         2
0  0.213845  0.499513  0.672325
1  0.607181  0.871924  0.765004
2  0.916804  0.050332  0.746759
3  0.834908  0.265275  0.791417
4  0.506703  0.310239  0.445191

In [401]:
df['mean'] = df.mean(axis=1)
df

0         1         2      mean
0  0.213845  0.499513  0.672325  0.461894
1  0.607181  0.871924  0.765004  0.748036
2  0.916804  0.050332  0.746759  0.571298
3  0.834908  0.265275  0.791417  0.630533
4  0.506703  0.310239  0.445191  0.420711

**24**. Suppose you have DataFrame with 10 columns of real numbers, for example:

    df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))
    Which column of numbers has the smallest sum? Return that column's label.

In [402]:
df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))
df

a         b         c         d         e         f         g  \
0  0.815734  0.397679  0.747455  0.242237  0.701322  0.230112  0.323743   
1  0.782680  0.818671  0.390797  0.312291  0.190944  0.365571  0.733129   
2  0.974679  0.106164  0.262214  0.626100  0.754874  0.460971  0.123858   
3  0.964612  0.415056  0.999828  0.796896  0.823251  0.841775  0.936806   
4  0.114349  0.152620  0.468728  0.871253  0.934595  0.822348  0.702534   

          h         i         j  
0  0.876458  0.201319  0.456506  
1  0.161711  0.518538  0.950876  
2  0.451539  0.451621  0.378194  
3  0.331561  0.898491  0.694399  
4  0.061339  0.340294  0.692580

In [463]:
df.sum(axis=0)

a    3.652054
b    1.890191
c    2.869022
d    2.848777
e    3.404986
f    2.720777
g    2.820071
h    1.882608
i    2.410262
j    3.172555
dtype: float64

In [466]:
df.sum(axis=0).idxmin()

'h'

**25**. How do you count how many unique rows a DataFrame has (i.e. ignore all rows that are duplicates)? As input, use a DataFrame of zeros and ones with 10 rows and 3 columns.

    df = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))

In [467]:
df = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))
df

0  1  2
0  1  0  0
1  1  1  0
2  0  1  1
3  0  0  0
4  1  0  1
5  0  1  1
6  1  1  0
7  1  0  0
8  1  0  1
9  1  1  0

In [476]:
df.value_counts()

0  1  2
1  1  0    3
0  1  1    2
1  0  0    2
      1    2
0  0  0    1
Name: count, dtype: int64

In [482]:
df.value_counts().count()

5

**26**. In the cell below, you have a DataFrame df that consists of 10 columns of floating-point numbers. Exactly 5 entries in each row are NaN values

    For each row of the DataFrame, find the column which contains the third NaN value.

    You should return a Series of column labels: e, c, d, h, d

In [3]:
nan = np.nan

data = [[0.04,  nan,  nan, 0.25,  nan, 0.43, 0.71, 0.51,  nan,  nan],
        [ nan,  nan,  nan, 0.04, 0.76,  nan,  nan, 0.67, 0.76, 0.16],
        [ nan,  nan, 0.5 ,  nan, 0.31, 0.4 ,  nan,  nan, 0.24, 0.01],
        [0.49,  nan,  nan, 0.62, 0.73, 0.26, 0.85,  nan,  nan,  nan],
        [ nan,  nan, 0.41,  nan, 0.05,  nan, 0.61,  nan, 0.48, 0.68]]

columns = list('abcdefghij')

df = pd.DataFrame(data, columns=columns)
df

a   b     c     d     e     f     g     h     i     j
0  0.04 NaN   NaN  0.25   NaN  0.43  0.71  0.51   NaN   NaN
1   NaN NaN   NaN  0.04  0.76   NaN   NaN  0.67  0.76  0.16
2   NaN NaN  0.50   NaN  0.31  0.40   NaN   NaN  0.24  0.01
3  0.49 NaN   NaN  0.62  0.73  0.26  0.85   NaN   NaN   NaN
4   NaN NaN  0.41   NaN  0.05   NaN  0.61   NaN  0.48  0.68

In [35]:
df.isna()

a     b      c      d      e      f      g      h      i      j
0  False  True   True  False   True  False  False  False   True   True
1   True  True   True  False  False   True   True  False  False  False
2   True  True  False   True  False  False   True   True  False  False
3  False  True   True  False  False  False  False   True   True   True
4   True  True  False   True  False   True  False   True  False  False

In [36]:
df.isna().cumsum(axis=1)

a  b  c  d  e  f  g  h  i  j
0  0  1  2  2  3  3  3  3  4  5
1  1  2  3  3  3  4  5  5  5  5
2  1  2  2  3  3  3  4  5  5  5
3  0  1  2  2  2  2  2  3  4  5
4  1  2  2  3  3  4  4  5  5  5

In [37]:
df.isna().cumsum(axis=1) == 3

a      b      c      d      e      f      g      h      i      j
0  False  False  False  False   True   True   True   True  False  False
1  False  False   True   True   True  False  False  False  False  False
2  False  False  False   True   True   True  False  False  False  False
3  False  False  False  False  False  False  False   True  False  False
4  False  False  False   True   True  False  False  False  False  False

In [40]:
# idxmin 은 False 값을 가진 첫번째 컬럼 idxmax 는 True 값을 가진 첫번째 컬럼.
(df.isna().cumsum(axis=1) == 3).idxmax(axis=1)

0    e
1    c
2    d
3    h
4    d
dtype: object

**27**. A DataFrame has a column of groups 'grps' and and column of integer values 'vals':

    df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                       'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})
    For each group, find the sum of the three greatest values. You should end up with the answer as follows:

    grps
    a    409
    b    156
    c    345

In [41]:
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'),
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})

df

grps  vals
0     a    12
1     a   345
2     a     3
3     b     1
4     b    45
5     c    14
6     a     4
7     a    52
8     b    54
9     c    23
10    c   235
11    c    21
12    b    57
13    b     3
14    c    87

In [100]:
df.groupby('grps')['vals'].nlargest(3)

grps    
a     1     345
      7      52
      0      12
b     12     57
      8      54
      4      45
c     10    235
      14     87
      9      23
Name: vals, dtype: int64

In [101]:
df.groupby('grps')['vals'].nlargest(3).groupby('grps').sum()

grps
a    409
b    156
c    345
Name: vals, dtype: int64

**28**. The DataFrame df constructed below has two integer columns 'A' and 'B'. The values in 'A' are between 1 and 100 (inclusive).

    For each group of 10 consecutive integers in 'A' (i.e. (0, 10], (10, 20], ...), calculate the sum of the corresponding values in column 'B'.

    The answer should be a Series as follows:
            A
        (0, 10]      635
        (10, 20]     360
        (20, 30]     315
        (30, 40]     306
        (40, 50]     750
        (50, 60]     284
        (60, 70]     424
        (70, 80]     526
        (80, 90]     835
        (90, 100]    852

In [102]:
df = pd.DataFrame(np.random.RandomState(8765).randint(1, 101, size=(100, 2)), columns = ["A", "B"])
df

A   B
0   46  29
1   75  22
2   49  63
3   33  43
4   71  75
..  ..  ..
95  60  87
96  57  40
97  86  19
98  50  56
99  97  94

[100 rows x 2 columns]

In [104]:
np.arange(0, 101, 10)

array([  0,  10,  20,  30,  40,  50,  60,  70,  80,  90, 100])

In [103]:
pd.cut(df['A'], np.arange(0, 101, 10))

0      (40, 50]
1      (70, 80]
2      (40, 50]
3      (30, 40]
4      (70, 80]
        ...    
95     (50, 60]
96     (50, 60]
97     (80, 90]
98     (40, 50]
99    (90, 100]
Name: A, Length: 100, dtype: category
Categories (10, interval[int64, right]): [(0, 10] < (10, 20] < (20, 30] < (30, 40] ... (60, 70] < (70, 80] < (80, 90] < (90, 100]]

In [107]:
df.groupby(pd.cut(df['A'], np.arange(0, 101, 10)), observed=True)['B'].sum()

A
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
Name: B, dtype: int64

## DataFrames: harder problems

### These might require a bit of thinking outside the box...
    ...but all are solvable using just the usual pandas/NumPy methods (and so avoid using explicit for loops).
    
    Difficulty: hard

**29**. Consider a DataFrame df where there is an integer column 'X':

    df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})
    For each value, count the difference back to the previous zero (or the start of the Series, whichever is closer). These values should therefore be
    
    [1, 2, 0, 1, 2, 3, 4, 0, 1, 2]
    Make this a new column 'Y'.

In [124]:
df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})
df

X
0  7
1  2
2  0
3  3
4  4
5  2
6  5
7  0
8  3
9  4

In [141]:
(df['X'] == 0)

0    False
1    False
2     True
3    False
4    False
5    False
6    False
7     True
8    False
9    False
Name: X, dtype: bool

In [142]:
(df['X'] == 0).cumsum()

0    0
1    0
2    1
3    1
4    1
5    1
6    1
7    2
8    2
9    2
Name: X, dtype: int64

In [143]:
df.groupby((df['X'] == 0).cumsum()).cumcount()

0    0
1    1
2    0
3    1
4    2
5    3
6    4
7    0
8    1
9    2
dtype: int64

In [133]:
df['Y'] = df.groupby((df['X'] == 0).cumsum()).cumcount()
df

X  Y
0  7  0
1  2  1
2  0  0
3  3  1
4  4  2
5  2  3
6  5  4
7  0  0
8  3  1
9  4  2

In [136]:
first_zero_prev_idx = (df['X'] == 0).idxmax() -1
df.Y.loc[0:first_zero_prev_idx]

0    0
1    1
Name: Y, dtype: int64

In [138]:
update_value = df.Y.loc[0:first_zero_prev_idx] + 1
update_value

0    1
1    2
Name: Y, dtype: int64

In [139]:
df.update(update_value)
df

X  Y
0  7  1
1  2  2
2  0  0
3  3  1
4  4  2
5  2  3
6  5  4
7  0  0
8  3  1
9  4  2

**30**. Consider the DataFrame constructed below which contains rows and columns of numerical data.

    Create a list of the column-row index locations of the 3 largest values in this DataFrame. In this case, the answer should be:
        [(5, 7), (6, 4), (2, 5)]

In [4]:
rawDf = pd.DataFrame(np.random.RandomState(30).randint(1, 101, size=(8, 8)))
rawDf

0   1   2   3   4   5   6   7
0  38  38  46  46  13  24   3  54
1  18  47   4  42   8  66  50  46
2  62  36  19  19  77  17   7  63
3  28  47  46  65  63  12  16  24
4  14  51  34  56  29  59  92  79
5  58  76  96  45  38  76  58  40
6  10  34  48  40  37  23  41  26
7  55  70  91  27  79  92  20  31

In [5]:
idx_Y = rawDf.idxmax(axis=0)
idx_Y

0    2
1    5
2    5
3    3
4    7
5    7
6    4
7    4
dtype: int64

In [6]:
idx_X = rawDf.idxmax(axis=1)
idx_X

0    7
1    5
2    4
3    3
4    6
5    2
6    2
7    5
dtype: int64

In [7]:
pd.DataFrame({'Y':idx_Y, 'X':idx_X})

Y  X
0  2  7
1  5  5
2  5  4
3  3  3
4  7  6
5  7  2
6  4  2
7  4  5

In [8]:
def convXY2(x):
    return [( x.name,  x['Y']), (x['X'], x.name)]

In [9]:
xyList = pd.DataFrame({'Y':idx_Y, 'X':idx_X}).apply( convXY2, axis=1 ).sum()
pd.DataFrame( {'XY': xyList} )

XY
0   (0, 2)
1   (7, 0)
2   (1, 5)
3   (5, 1)
4   (2, 5)
5   (4, 2)
6   (3, 3)
7   (3, 3)
8   (4, 7)
9   (6, 4)
10  (5, 7)
11  (2, 5)
12  (6, 4)
13  (2, 6)
14  (7, 4)
15  (5, 7)

In [10]:
df = pd.DataFrame( {'XY': xyList} )
df

XY
0   (0, 2)
1   (7, 0)
2   (1, 5)
3   (5, 1)
4   (2, 5)
5   (4, 2)
6   (3, 3)
7   (3, 3)
8   (4, 7)
9   (6, 4)
10  (5, 7)
11  (2, 5)
12  (6, 4)
13  (2, 6)
14  (7, 4)
15  (5, 7)

In [11]:
df.drop_duplicates(inplace=True)
df

XY
0   (0, 2)
1   (7, 0)
2   (1, 5)
3   (5, 1)
4   (2, 5)
5   (4, 2)
6   (3, 3)
8   (4, 7)
9   (6, 4)
10  (5, 7)
13  (2, 6)
14  (7, 4)

In [12]:
df['value'] = df.XY.map( lambda x: rawDf[x[0]][x[1]] )
df

XY  value
0   (0, 2)     62
1   (7, 0)     54
2   (1, 5)     76
3   (5, 1)     66
4   (2, 5)     96
5   (4, 2)     77
6   (3, 3)     65
8   (4, 7)     79
9   (6, 4)     92
10  (5, 7)     92
13  (2, 6)     48
14  (7, 4)     79

In [13]:
df.nlargest(3, columns='value')

XY  value
4   (2, 5)     96
9   (6, 4)     92
10  (5, 7)     92

**31**. You are given the DataFrame below with a column of group IDs, 'grps', and a column of corresponding integer values, 'vals'.

    df = pd.DataFrame({"vals": np.random.RandomState(31).randint(-30, 30, size=15),
                       "grps": np.random.RandomState(31).choice(["A", "B"], 15)})
    Create a new column 'patched_values' which contains the same values as the 'vals' any negative values in 'vals' with the group mean:

            vals grps  patched_vals
        0    -12    A          13.6
        1     -7    B          28.0
        2    -14    A          13.6
        3      4    A           4.0
        4     -7    A          13.6
        5     28    B          28.0
        6     -2    A          13.6
        7     -1    A          13.6
        8      8    A           8.0
        9     -2    B          28.0
        10    28    A          28.0
        11    12    A          12.0
        12    16    A          16.0
        13   -24    A          13.6
        14   -12    A          13.6

In [158]:
df = pd.DataFrame({"vals": np.random.RandomState(31).randint(-30, 30, size=15),
    "grps": np.random.RandomState(31).choice(["A", "B"], 15)})

df

vals grps
0    -12    A
1     -7    B
2    -14    A
3      4    A
4     -7    A
5     28    B
6     -2    A
7     -1    A
8      8    A
9     -2    B
10    28    A
11    12    A
12    16    A
13   -24    A
14   -12    A

In [155]:
mean_value = df[ df.vals > -1].groupby('grps').mean()
mean_value

vals
grps      
A     13.6
B     28.0

In [157]:
df['patched_values'] = df.apply(lambda x: mean_value.vals[x.grps] if x.vals < 0 else x.vals, axis=1 )
df

vals grps  patched_values
0    -12    A            13.6
1     -7    B            28.0
2    -14    A            13.6
3      4    A             4.0
4     -7    A            13.6
5     28    B            28.0
6     -2    A            13.6
7     -1    A            13.6
8      8    A             8.0
9     -2    B            28.0
10    28    A            28.0
11    12    A            12.0
12    16    A            16.0
13   -24    A            13.6
14   -12    A            13.6

**32**. Implement a rolling mean over groups with window size 3, which ignores NaN value. For example consider the following DataFrame:



    df = pd.DataFrame({'group': list('aabbabbbabab'),
                           'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})
    df
       group  value
    0      a    1.0
    1      a    2.0
    2      b    3.0
    3      b    NaN
    4      a    2.0
    5      b    3.0
    6      b    NaN
    7      b    1.0
    8      a    7.0
    9      b    3.0
    10     a    NaN
    11     b    8.0

    The goal is to compute the Series:

        0     1.000000
        1     1.500000
        2     3.000000
        3     3.000000
        4     1.666667
        5     3.000000
        6     3.000000
        7     2.000000
        8     3.666667
        9     2.000000
        10    4.500000
        11    4.000000

    E.g. the first window of size three for group 'b' has values 3.0, NaN and 3.0 and occurs at row index 5. Instead of being NaN the value in the new column at this row index should be 3.0 (just the two non-NaN values are used to compute the mean (3+3)/2)

In [258]:
df = pd.DataFrame({'group': list('aabbabbbabab'),
                    'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})
df

group  value
0      a    1.0
1      a    2.0
2      b    3.0
3      b    NaN
4      a    2.0
5      b    3.0
6      b    NaN
7      b    1.0
8      a    7.0
9      b    3.0
10     a    NaN
11     b    8.0

In [259]:
df.reset_index()

index group  value
0       0     a    1.0
1       1     a    2.0
2       2     b    3.0
3       3     b    NaN
4       4     a    2.0
5       5     b    3.0
6       6     b    NaN
7       7     b    1.0
8       8     a    7.0
9       9     b    3.0
10     10     a    NaN
11     11     b    8.0

In [298]:
def convMean(x):
    divideCnt = (0 if x['current'] == 0 else 1) + (0 if x['prev_1'] == 0 else 1) + (0 if x['prev_2'] == 0 else 1)
    return  (x['current'] + x['prev_1'] + x['prev_2']) / divideCnt

In [300]:
df_a =  df.loc[df.group == 'a'].pivot(columns='group', values='value')
df_a.rename(columns={'a':'current'}, inplace=True)
df_a['prev_1'] = df_a.shift(1)['current']
df_a['prev_2'] = df_a.shift(2)['current']
df_a.fillna(0, inplace=True)
df_a['mean'] = df_a.apply( convMean  , axis=1)
df_a

group  current  prev_1  prev_2      mean
0          1.0     0.0     0.0  1.000000
1          2.0     1.0     0.0  1.500000
4          2.0     2.0     1.0  1.666667
8          7.0     2.0     2.0  3.666667
10         0.0     7.0     2.0  4.500000

In [301]:
df_b = df.loc[df.group == 'b'].pivot(columns='group', values='value')
df_b.rename(columns={'b':'current'}, inplace=True)
df_b['prev_1'] = df_b.shift(1)['current']
df_b['prev_2'] = df_b.shift(2)['current']
df_b.fillna(0, inplace=True)
df_b['mean'] = df_b.apply( convMean  , axis=1)
df_b

group  current  prev_1  prev_2  mean
2          3.0     0.0     0.0   3.0
3          0.0     3.0     0.0   3.0
5          3.0     0.0     3.0   3.0
6          0.0     3.0     0.0   3.0
7          1.0     0.0     3.0   2.0
9          3.0     1.0     0.0   2.0
11         8.0     3.0     1.0   4.0

In [314]:
result = pd.concat([df_a, df_b]).sort_index()['mean'].map(lambda x:  '%.6f' % x )
result

0     1.000000
1     1.500000
2     3.000000
3     3.000000
4     1.666667
5     3.000000
6     3.000000
7     2.000000
8     3.666667
9     2.000000
10    4.500000
11    4.000000
Name: mean, dtype: object

## Series and DatetimeIndex

### Exercises for creating and manipulating Series with datetime data
Difficulty: easy/medium

    pandas is fantastic for working with dates and times. These puzzles explore some of this functionality.

**33** . Create a DatetimeIndex that contains each business day of 2015 and use it to index a Series of random numbers. Let's call this Series s.

In [320]:
idx_date =pd.date_range(start='2015-01-01', end='2015-12-31', freq='B')
idx_date

DatetimeIndex(['2015-01-01', '2015-01-02', '2015-01-05', '2015-01-06',
               '2015-01-07', '2015-01-08', '2015-01-09', '2015-01-12',
               '2015-01-13', '2015-01-14',
               ...
               '2015-12-18', '2015-12-21', '2015-12-22', '2015-12-23',
               '2015-12-24', '2015-12-25', '2015-12-28', '2015-12-29',
               '2015-12-30', '2015-12-31'],
              dtype='datetime64[ns]', length=261, freq='B')

In [323]:
idx_date.size

261

In [350]:
s = pd.Series(np.random.randint(0, 100, idx_date.size), idx_date)
s

2015-01-01    28
2015-01-02    21
2015-01-05    40
2015-01-06    65
2015-01-07    79
              ..
2015-12-25    50
2015-12-28     1
2015-12-29    27
2015-12-30     4
2015-12-31    12
Freq: B, Length: 261, dtype: int64

**34**. Find the sum of the values in s for every Wednesday.

In [351]:
idx_date.day_name()

Index(['Thursday', 'Friday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday',
       'Friday', 'Monday', 'Tuesday', 'Wednesday',
       ...
       'Friday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday',
       'Monday', 'Tuesday', 'Wednesday', 'Thursday'],
      dtype='object', length=261)

In [358]:
s.loc[s.index.day_name() == 'Wednesday'].sum().item()

2769

**35**. For each calendar month in s, find the mean of values.

In [365]:
s.groupby(idx_date.month).mean().map(lambda x:  '%.2f' % x )

1     46.32
2     44.10
3     49.36
4     49.23
5     66.48
6     52.00
7     51.65
8     55.90
9     42.82
10    32.77
11    47.19
12    35.83
dtype: object

**36**. For each group of four consecutive calendar months in s, find the date on which the highest value occurred.

In [459]:
df = pd.DataFrame(s, index=idx_date, columns=['value'])
df.head()

value
2015-01-01    28
2015-01-02    21
2015-01-05    40
2015-01-06    65
2015-01-07    79

In [460]:
df['month_boundary'] =  df.apply( lambda x : pd.cut(x.index.month, bins = [0,4,8, 12], labels=['[1..4]', '[5..8]', '[9..12]']) )
df.head()

value month_boundary
2015-01-01    28         [1..4]
2015-01-02    21         [1..4]
2015-01-05    40         [1..4]
2015-01-06    65         [1..4]
2015-01-07    79         [1..4]

In [470]:
df.groupby('month_boundary', observed=True).agg(
    date= pd.NamedAgg(column="value", aggfunc="idxmax"),
    value= pd.NamedAgg(column="value", aggfunc="max")
)

date value
month_boundary                 
[1..4]         2015-02-05    97
[5..8]         2015-08-12    99
[9..12]        2015-09-10    98

**37**. Create a DateTimeIndex consisting of the third Thursday in each month for the years 2015 and 2016.